### DISCUSSION TITLE
STIMER functionality with no 32kHz crystal installed

### DISCUSSION THREAD

**Discussion by breadboardbasics** [AmbiqMicro/ambiqzephyr#21] [Q&A] [answered]
Hello!

I am working on a project where we are using an Apollo3 for signal processing purposes in a size constrained area. We have been modifying the existing EVB board in Zephyr in this initial development phase we are in. Due to the size concerns, we are not designing in a 32.768kHz crystal. From initial testing, there has been some strange behavior when looking at the tick count as well as k_msleep causing freezing. 

Digging into the datasheet as well as some of the underlying Zephyr code, I believe that I have traced down the issue to the fact that the STIMER is being used for system clock purposes by default and the STIMER source is being set as the XT oscillator which of course uses the external 32kHz crystal which we don't have installed. 

I have tried a lot of modifications to get things working properly without the crystal but have not been having success thus far. I was wondering if anyone here has any advice on what to change and how to set things up properly. 

Overall, I think that it is important to add a system clock source which doesn't rely on the XT oscillator being on constantly since it is power hungry and will cause unnecessary power draw when the CPU is idle. 

Excited for all the work that is being done to add more support for Zephyr from the Ambiq team!

**Comment by AlessandroLuo** [comment, marked as answer]
Hi @breadboardbasics , thanks for your comment, you can change the AM_HAL_STIMER_XTAL_32KHZ macro in stimer_init to other options for example AM_HAL_STIMER_HFRC_187_5KHZ, it will use devided HFRC as clock source, and in the meantime you need to modify SYS_CLOCK_HW_CYCLES_PER_SEC accordingly.
It's a very good catch for optimization, we will add this to our plan to make the stimer clock source configurable.

**Reply by breadboardbasics** [reply]
Thank you! That is working on our custom board.

**Reply by AlessandroLuo** [reply]
Great to hear that! there's already a PR https://github.com/zephyrproject-rtos/zephyr/pull/81852 on the way for this change, please rebase to the latest main once it's merged.
